homeassistant/components/edl21/sensor.py

```diff
@@ -0,0 +1,164 @@
+    }
+
+    def __init__(self, hass, config, async_add_entities) -> None:
+        """Initialize an EDL21 object."""
+        self._blacklist = {
+            # A=129: Manufacturer specific
+            "129-129:199.130.3*255",  # Iskraemeco: Manufacturer
+            "129-129:199.130.5*255",  # Iskraemeco: Public Key
+        }
+        self._cache = {}
+        self._hass = hass
+        self._async_add_entities = async_add_entities
+        self._proto = SmlProtocol(config[CONF_SERIAL_PORT])
+        self._proto.add_listener(self.event, ["SmlGetListResponse"])
+
+    async def connect(self):
+        """Connect to an EDL21 reader."""
+        await self._proto.connect(self._hass.loop)
+
+    def event(self, message_body) -> None:
+        """Handle events from pysml."""
+        assert isinstance(message_body, SmlGetListResponse)
+
+        new_devices = []
+        for telegram in message_body.get("valList", []):
+            obis = telegram.get("objName")
+            if not obis:
+                continue
+
+            device = self._cache.get(obis)
+            if not device:
+                name = self._OBIS_NAMES.get(obis)
+                if name:
+                    device = self._cache[obis] = EDL21Entity(obis, name, telegram)
+                    new_devices.append(device)
+                elif obis not in self._blacklist:
+                    _LOGGER.warning(
+                        "Unhandled sensor %s detected. Please report at "
+                        "https://github.com/home-assistant/home-assistant/issues",
+                        obis,
+                    )
+                    self._blacklist.add(obis)
+            elif device.update_telegram(telegram):
+                self._hass.async_create_task(device.async_update_ha_state())
+
+        if new_devices:
+            self._hass.async_add_job(
+                self._async_add_entities(new_devices, update_before_add=True)
+            )
+
+
+class EDL21Entity(Entity):
+    """Entity reading values from EDL21 telegram."""
+
+    def __init__(self, obis, name, telegram):
+        """Initialize an EDL21Entity."""
+        self._obis = obis
+        self._name = name
+        self._telegram = telegram
+
+    @property
+    def should_poll(self) -> bool:
+        """Do not poll."""
+        return False
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self._obis
+
+    @property
+    def name(self) -> Optional[str]:
+        """Return a name."""
+        return self._name
+
+    @property
+    def state(self) -> str:
+        """Return the value of the last received telegram."""
+        return self._telegram.get("value")
+
+    @property
+    def device_state_attributes(self):
+        """Enumerate supported attributes."""
+        attr = self._telegram.copy()
+        for key in ("objName", "unit", "value"):
+            if key in attr:
+                del attr[key]
+        return attr
```

Comment:
Please write out the attribute keys that are in use here. We want to check what they are and the formatting.